azurerm/resource_arm_virtual_machine.go

```diff
@@ -1598,3 +1597,91 @@ func resourceArmVirtualMachineStorageImageReferenceHash(v interface{}) int {
 	}
 	return hashcode.String(buf.String())
 }
+
+func resourceArmVirtualMachineReviseStorageInfo(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).vmClient
+	ctx := meta.(*ArmClient).StopContext
+
+	id, err := parseAzureResourceID(d.Id())
+	if err != nil {
+		return err
+	}
+	resGroup := id.ResourceGroup
+	name := id.Path["virtualMachines"]
+
+	instance, err := client.InstanceView(ctx, resGroup, name)
+	if err != nil {
+		return fmt.Errorf("Error making InstanceView request on Azure Virtual Machine %s: %+v", name, err)
+	}
+
+	stopped := false
+	if instance.Statuses != nil {
+		for _, status := range *instance.Statuses {
+			if status.Code != nil &&
+				(*status.Code == "PowerState/deallocated" ||
+					*status.Code == "PowerState/deallocating" ||
+					*status.Code == "PowerState/starting" ||
+					*status.Code == "PowerState/stopping" ||
+					*status.Code == "PowerState/stopped") {
+				stopped = true
+				break
+			}
+		}
+	}
+
+	if stopped {
+		if osDisks, ok := d.GetOk("storage_os_disk"); ok {
+			if err := resourceArmVirtualMachineUpdateManagedDisk(osDisks.([]interface{}), meta); err != nil {
+				return err
+			}
+			if err := d.Set("storage_os_disk", osDisks); err != nil {
+				return fmt.Errorf("[DEBUG] Error setting virtual machine revised storage OS disk error: %#v", err)
+			}
+		}
+		if dataDisks, ok := d.GetOk("storage_data_disk"); ok {
+			if err := resourceArmVirtualMachineUpdateManagedDisk(dataDisks.([]interface{}), meta); err != nil {
```

Comment:
updating the state twice isn't ideal here - we should assign this to the objects used prior to the flatten methods e.g. https://github.com/terraform-providers/terraform-provider-azurerm/blob/master/azurerm/resource_arm_virtual_machine.go#L960